### PR TITLE
OHRM5X-1683: Develop dashboard - employee time at work API

### PR DESCRIPTION
**Pull request by SandeepaOHRM**

### CODE REVIEW

**Review by RajithaKumara — CHANGES_REQUESTED** [review]

File: src/plugins/orangehrmDashboardPlugin/config/routes.yaml [hunk 1]
```diff
@@ -26,6 +26,14 @@ apiv2_dashboard_employees_by_location:
   defaults:
     _api: OrangeHRM\Dashboard\Api\EmployeeDistributionByLocationAPI
 
+apiv2_dashboard_employee_time_at_work:
+  path: /api/v2/dashboard/employees/{empNumber}/time-at-work
```
Comment (RajithaKumara) [inline src/plugins/orangehrmDashboardPlugin/config/routes.yaml:30]:
```suggestion
  path: /api/v2/dashboard/employees/time-at-work
```
will move `empNumber` to query param

File: src/plugins/orangehrmDashboardPlugin/config/routes.yaml [hunk 2]
```diff
@@ -26,6 +26,14 @@ apiv2_dashboard_employees_by_location:
...
+  controller: OrangeHRM\Core\Controller\Rest\V2\GenericRestController::handle
+  methods: [ GET ]
+  defaults:
+    _api: OrangeHRM\Dashboard\Api\EmployeeTimeAtWorkAPI
+    _key: empNumber
```
Comment (RajithaKumara) [inline src/plugins/orangehrmDashboardPlugin/config/routes.yaml:35]:
```suggestion
    id: 0
```
to go to getOne

File: src/plugins/orangehrmDashboardPlugin/Api/EmployeeTimeAtWorkAPI.php [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+use OrangeHRM\Core\Api\V2\Validator\Rules;
+use OrangeHRM\Core\Traits\Auth\AuthUserTrait;
+use OrangeHRM\Dashboard\Traits\Service\EmployeeTimeAtWorkServiceTrait;
+
+class EmployeeTimeAtWorkAPI extends Endpoint implements ResourceEndpoint
+{
+    use AuthUserTrait;
+    use EmployeeTimeAtWorkServiceTrait;
+
+    public const PARAMETER_EMP_NUMBER = 'empNumber';
```
[the remark this answers is not in the document]
Comment (SandeepaOHRM) [reply src/plugins/orangehrmDashboardPlugin/Api/EmployeeTimeAtWorkAPI.php]: done

File: src/plugins/orangehrmDashboardPlugin/Api/EmployeeTimeAtWorkAPI.php [hunk 4]
```diff
@@ -0,0 +1,104 @@
...
+
+    /**
+     * @inheritDoc
+     */
+    public function getOne(): EndpointResult
+    {
+        $empNumber = $this->getRequestParams()->getInt(
+            RequestParams::PARAM_TYPE_ATTRIBUTE,
+            self::PARAMETER_EMP_NUMBER
```
[the remark this answers is not in the document]
Comment (SandeepaOHRM) [reply src/plugins/orangehrmDashboardPlugin/Api/EmployeeTimeAtWorkAPI.php]: done

File: src/plugins/orangehrmDashboardPlugin/Api/EmployeeTimeAtWorkAPI.php [hunk 5]
```diff
@@ -0,0 +1,104 @@
...
+    }
+
+    /**
+     * @inheritDoc
+     */
+    public function getValidationRuleForGetOne(): ParamRuleCollection
+    {
+        return new ParamRuleCollection(
+            new ParamRule(
+                self::PARAMETER_EMP_NUMBER,
```
[the remark this answers is not in the document]
Comment (SandeepaOHRM) [reply src/plugins/orangehrmDashboardPlugin/Api/EmployeeTimeAtWorkAPI.php]: done

File: src/plugins/orangehrmDashboardPlugin/Dao/EmployeeTimeAtWorkDao.php [hunk 6]
```diff
@@ -0,0 +1,70 @@
...
+     * @param int $empNumber
+     * @return AttendanceRecord|null
+     */
+    public function getLatestAttendanceRecordByEmpNumber(int $empNumber): ?AttendanceRecord
+    {
+        return $this->getRepository(AttendanceRecord::class)
+            ->findOneBy(
+                ['employee' => $empNumber],
+                ['id' => ListSorter::DESCENDING]
+            );
```
[the remark this answers is not in the document]
Comment (SandeepaOHRM) [reply src/plugins/orangehrmDashboardPlugin/Dao/EmployeeTimeAtWorkDao.php:39]: done

File: src/plugins/orangehrmDashboardPlugin/Dao/EmployeeTimeAtWorkDao.php [hunk 7]
```diff
@@ -0,0 +1,70 @@
...
+                $qb->expr()->between(
+                    'attendanceRecord.punchInUserTime',
+                    ':start',
+                    ':end'
+                ),
+                $qb->expr()->between(
+                    'attendanceRecord.punchOutUserTime',
+                    ':start',
+                    ':end'
+                ),
```
[the remark this answers is not in the document]
Comment (SandeepaOHRM) [reply src/plugins/orangehrmDashboardPlugin/Dao/EmployeeTimeAtWorkDao.php]: yes

File: src/plugins/orangehrmDashboardPlugin/Service/EmployeeTimeAtWorkService.php [hunk 8]
```diff
@@ -0,0 +1,275 @@
...
+    }
+
+    /**
+     * @param int $empNumber
+     * @return array
+     * @throws Exception
+     */
+    public function getTimeAtWorkMetaData(int $empNumber): array
+    {
+        //TODO::Determine whether current date is server date or date from the client side
```
[the remark this answers is not in the document]
Comment (SandeepaOHRM) [reply src/plugins/orangehrmDashboardPlugin/Service/EmployeeTimeAtWorkService.php]: done

File: src/plugins/orangehrmDashboardPlugin/config/routes.yaml (hunk 1 above)
Comment (SandeepaOHRM) [reply src/plugins/orangehrmDashboardPlugin/config/routes.yaml]: done

File: src/plugins/orangehrmDashboardPlugin/config/routes.yaml (hunk 2 above)
Comment (SandeepaOHRM) [reply src/plugins/orangehrmDashboardPlugin/config/routes.yaml]: done

File: src/plugins/orangehrmDashboardPlugin/Dao/EmployeeTimeAtWorkDao.php [hunk 9]
```diff
@@ -0,0 +1,78 @@
...
+    /**
+     * @param int $empNumber
+     * @return AttendanceRecord|null
+     */
+    public function getLatestAttendanceRecordByEmpNumber(int $empNumber): ?AttendanceRecord
+    {
+        return $this->getRepository(AttendanceRecord::class)
+            ->findOneBy(
+                ['employee' => $empNumber],
+                ['id' => ListSorter::DESCENDING]
```
Comment (RajithaKumara) [inline src/plugins/orangehrmDashboardPlugin/Dao/EmployeeTimeAtWorkDao.php:38]: check how sorted in attendance module also. want to check is date also involved

File: src/plugins/orangehrmDashboardPlugin/Dao/EmployeeTimeAtWorkDao.php [hunk 10]
```diff
@@ -0,0 +1,78 @@
...
+                $qb->expr()->between(
+                    'attendanceRecord.punchOutUtcTime',
+                    ':start',
+                    ':end'
+                ),
+                $qb->expr()->andX(
+                    $qb->expr()->lte('attendanceRecord.punchInUtcTime', ':start'),
+                    $qb->expr()->gte('attendanceRecord.punchOutUtcTime', ':end')
+                )
+            )
```
Comment (RajithaKumara) [inline src/plugins/orangehrmDashboardPlugin/Dao/EmployeeTimeAtWorkDao.php:72]: hope all scenarios covered via tests

File: src/plugins/orangehrmDashboardPlugin/Service/EmployeeTimeAtWorkService.php [hunk 11]
```diff
@@ -0,0 +1,275 @@
...
+
+    /**
+     * @var TimesheetPeriodService
+     */
+    private TimesheetPeriodService $timesheetPeriodService;
+
+    /**
+     * @var int
+     */
+    private int $totalTimeForWeek = 0;
```
[the remark this answers is not in the document]
Comment (RajithaKumara) [reply src/plugins/orangehrmDashboardPlugin/Service/EmployeeTimeAtWorkService.php:55]: check this if not refactored yet

File: src/plugins/orangehrmDashboardPlugin/Service/EmployeeTimeAtWorkService.php [hunk 12]
```diff
@@ -0,0 +1,287 @@
...
+
+    public function getEmployeeTimeAtWorkDao(): EmployeeTimeAtWorkDao
+    {
+        return $this->employeeTimeAtWorkDao ??= new EmployeeTimeAtWorkDao();
+    }
+
+    /**
+     * @return TimesheetPeriodService
+     */
+    private function getTimesheetPeriodService(): TimesheetPeriodService
```
Comment (RajithaKumara) [inline src/plugins/orangehrmDashboardPlugin/Service/EmployeeTimeAtWorkService.php:65]: `private`, above method `public`
it's okay to keep getters public. or make all private if there is no special reason

File: src/plugins/orangehrmDashboardPlugin/Service/EmployeeTimeAtWorkService.php [hunk 13]
```diff
@@ -0,0 +1,287 @@
...
+            }
+        }
+        return $totalTime;
+    }
+
+    /**
+     * @param DateTime $date
+     * @return array  eg:- array(if monday as first day in config => '2021-12-13', '2021-12-19')
+     */
+    private function getWeekBoundaryForGivenDate(DateTime $date): array
```
Comment (RajithaKumara) [inline src/plugins/orangehrmDashboardPlugin/Service/EmployeeTimeAtWorkService.php:218]: will reuse the attendance module method if possible. if this method not same as attendance, no need changes

File: src/plugins/orangehrmDashboardPlugin/Service/EmployeeTimeAtWorkService.php [hunk 14]
```diff
@@ -0,0 +1,287 @@
...
+                    'date' => $this->getDateTimeHelper()->formatDateTimeToYmd($startDateTime),
+                ],
+                'totalTime' => $this->getTimeInHoursAndMinutes($totalTimeForDay),
+            ];
+            $startUTCDateTime = clone $startUTCDateTime;
+            $startUTCDateTime = $startUTCDateTime->add(new DateInterval('P1D'));
+            $startDateTime = clone $startDateTime;
+            $startDateTime = $startDateTime->add(new DateInterval('P1D'));
+
+            $this->totalTimeForWeek = $this->totalTimeForWeek + $totalTimeForDay;
```
Comment (RajithaKumara) [inline src/plugins/orangehrmDashboardPlugin/Service/EmployeeTimeAtWorkService.php:261]:
```suggestion
            $totalTimeForWeek += $totalTimeForDay;
```
```php
    /**
     * @param int $empNumber
     * @param DateTime $currentDateTime
     * @return array
     * @throws Exception
     */
    private function getDataForCurrentWeek(
        int $empNumber,
        DateTime $currentDateTime
    ): array {
        // ...
        $totalTimeForWeek = 0;
        while ($counter < 7) {
            //...
            $totalTimeForWeek +=  $totalTimeForDay;
            $counter++;
        }
        return [$weeklyData, $totalTimeForWeek];
    }
```
if you want to keep `$this->totalTimeForWeek` you can expose only one public method to call from API class.

```php
        list($data, $meta) = $this->getEmployeeTimeAtWorkService()->getTimeAtWork();
        return new EndpointResourceResult(
            ArrayModel::class,
            $data,
            new ParameterBag($meta)
        );
```
```php
    public function getTimeAtWork(/** params */)
    {
        return [
            $this->getEmployeeTimeAtWorkService()->getTimeAtWorkData($empNumber, $currentDateTime),
            $this->getEmployeeTimeAtWorkService()->getTimeAtWorkMetaData($empNumber, $currentDateTime),
    }

    private function getTimeAtWorkData(/** params */) 

    private function getTimeAtWorkMetaData(/** params */) 
```
the reason, why I request to change, other devs don't know they should call `getTimeAtWorkMetaData` after `getTimeAtWorkData` to get correct result.

File: src/plugins/orangehrmDashboardPlugin/Service/EmployeeTimeAtWorkService.php [hunk 15]
```diff
@@ -0,0 +1,287 @@
...
+        return [
+            'lastAction' => $this->getLastActionDetails($empNumber),
+            'currentDay' => [
+                'currentDate' => $this->getDateDetails($currentDateTime),
+                'totalTime' => $this->getTimeInHoursAndMinutes($totalTimeForCurrentDay)
+            ],
+            'currentWeek' => [
+                'startDate' => $this->getDateDetails($weekStartDate),
+                'endDate' => $this->getDateDetails($weekEndDate),
+                'totalTime' => $this->getTimeInHoursAndMinutes($this->totalTimeForWeek)
```
Comment (RajithaKumara) [inline src/plugins/orangehrmDashboardPlugin/Service/EmployeeTimeAtWorkService.php:107]:
```suggestion
                'totalTime' => $this->getTimeInHoursAndMinutes($totalTimeForWeek)
```
```php
    public function getTimeAtWorkMetaData(int $empNumber, DateTime $currentDateTime, int $totalTimeForWeek): array
    {
        //...
        return [
            //...
            'currentWeek' => [
               //...
                'totalTime' => $this->getTimeInHoursAndMinutes($totalTimeForWeek)
            ]
        ];
    }
```

File: src/plugins/orangehrmDashboardPlugin/test/Dao/EmployeeTimeAtWorkDaoTest.php [hunk 16]
```diff
@@ -0,0 +1,91 @@
...
+    protected function setUp(): void
+    {
+        $this->employeeTimeAtWorkDao = new EmployeeTimeAtWorkDao();
+        $this->fixture = Config::get(
+            Config::PLUGINS_DIR
+        ) . '/orangehrmDashboardPlugin/test/fixtures/EmployeeTimeAtWork.yaml';
+        TestDataService::populate($this->fixture);
+    }
+
+    public function testGetLatestAttendanceRecordByEmpNumber()
```
Comment (RajithaKumara) [inline src/plugins/orangehrmDashboardPlugin/test/Dao/EmployeeTimeAtWorkDaoTest.php:55]:
```suggestion
    public function testGetLatestAttendanceRecordByEmpNumber(): void
```

File: src/plugins/orangehrmDashboardPlugin/test/Dao/EmployeeTimeAtWorkDaoTest.php [hunk 17]
```diff
@@ -0,0 +1,91 @@
...
+        $this->assertEquals('5.5', $lastAttendanceRecord->getPunchOutTimeOffset());
+
+        /**
+         * No last attendance record found
+         */
+        $lastAttendanceRecord = $this->employeeTimeAtWorkDao->getLatestAttendanceRecordByEmpNumber(6);
+        $this->assertNotInstanceOf(AttendanceRecord::class, $lastAttendanceRecord);
+    }
+
+    public function testGetAttendanceRecordsByEmployeeAndDate()
```
Comment (RajithaKumara) [inline src/plugins/orangehrmDashboardPlugin/test/Dao/EmployeeTimeAtWorkDaoTest.php:73]:
```suggestion
    public function testGetAttendanceRecordsByEmployeeAndDate(): void
```

File: src/plugins/orangehrmDashboardPlugin/test/fixtures/QuickLaunchAPITest3.yml [hunk 18]
```diff
@@ -20,7 +20,11 @@ UserRole:
 User:
   - {"id": 1,"user_role_id": 1,"emp_number": 1,"user_name": "admin","user_password": "$2y$12$cRXToxVun41SeRgsWTSzP.2cJmj4AZWi/0.BZr529V8p68PCSnRQ."}
   - {"id": 2,"user_role_id": 2,"emp_number": 2,"user_name": "sharuka","user_password": "$2y$12$EH.mp1hqHt7BW4EpaeOCKeQTyokMLYWVpMSitHtoMu/W3K2xOf5p."}
-  - {"id": 3,"user_role_id": 2,"emp_number": 3,"user_name": "rajitha","user_password": "$2y$12$f2zH9NNS.GrX/Dmx8r2eWuJt986aPLOUF8LvEj.8aAvFxEmaPIoNa"}
+  - {"id": 3,"user_role_id": 3,"emp_number": 3,"user_name": "rajitha","user_password": "$2y$12$f2zH9NNS.GrX/Dmx8r2eWuJt986aPLOUF8LvEj.8aAvFxEmaPIoNa"}
```
Comment (RajithaKumara) [inline src/plugins/orangehrmDashboardPlugin/test/fixtures/QuickLaunchAPITest3.yml:23]: Btw, I don't like to use my name.
use https://github.com/orangehrm/orangehrm/blob/main/src/plugins/orangehrmLeavePlugin/test/fixtures/EmployeeLeaveEntitlementUsageLeaveReportDataAPITest.yaml#L59-L65 kind of dummy data :relaxed:

**Review by RajithaKumara — APPROVED** [review]

**Pull request merged**